csharp/src/Ice/InputStream.cs

```diff
@@ -886,18 +913,28 @@ internal string ReadFacet()
         /// <returns>The size read from the stream.</returns>
         internal int ReadSize()
         {
-            byte b = ReadByte();
-            if (b < 255)
+            if (OldEncoding)
             {
-                return b;
-            }
+                byte b = ReadByte();
+                if (b < 255)
+                {
+                    return b;
+                }
 
-            int size = ReadInt();
-            if (size < 0)
+                int size = ReadInt();
+                if (size < 0)
+                {
+                    throw new InvalidDataException($"read invalid size: {size}");
+                }
+                return size;
+            }
+            else
             {
-                throw new InvalidDataException($"read invalid size: {size}");
+                checked
+                {
+                    return (int)ReadVarUInt();
```

Comment:
Use `ReadVarULong()` instead to be consistent with the 2.0 size encoding? We have 2 implementations of ReadSize, should we perhaps call the static method here instead?

Reply:
Yes, fixed.

Reply:
Note the encoding for varulong and varuint is identical. varuint (and varint) are purely about language mapping. OutputStream's WriteSize actually uses:
```
    WriteVarUInt((uint)v);
```
where v is an int. The resulting encoding uses 1, 2, 4 or 8 bytes.

Reply:
Right. I actually missed the write, I guess it should use WriteVarULong as well :)? I don't see why we'd go through the hoops of Write/ReadVarUInt.

Reply:
Fixed as well.